<a href="https://colab.research.google.com/github/Jomifum/Assignment7D602/blob/main/Copy_of_07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**
#By Jose Fuentes

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include:

* Summary statistics means, medians, quartiles,
* Missing value information
* Any other relevant information about the dataset.  

Steps done in this sections described below:

Loading the Dataset: The dataset was loaded directly from the provided GitHub link using pandas.read_csv().

Summary Statistics performed: Summary statistics, including mean, median, and quartiles for numerical columns, were computed using describe() to understand the central tendency and distribution.

Missing Value Information: The amount of missing data in each column was assessed using isnull().sum(), which helped identify columns that require cleaning or imputation.


In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/Jomifum/Assignment7D602/main/Air_Quality.csv"
dataset = pd.read_csv(url)

# Display the first few rows of the dataset
print("Original Dataset:")
print(dataset.head(10))

# Summary statistics
summary_stats = dataset.describe(include='all')
print("\nSummary Statistics:")
print(summary_stats)

# Missing value information
missing_values = dataset.isnull().sum()
print("\nMissing Values Information:")
print(missing_values)

Original Dataset:
   Unique ID  Indicator ID                                   Name  \
0     179772           640  Boiler Emissions- Total SO2 Emissions   
1     179785           640  Boiler Emissions- Total SO2 Emissions   
2     178540           365                Fine particles (PM 2.5)   
3     178561           365                Fine particles (PM 2.5)   
4     823217           365                Fine particles (PM 2.5)   
5     177910           365                Fine particles (PM 2.5)   
6     177952           365                Fine particles (PM 2.5)   
7     177973           365                Fine particles (PM 2.5)   
8     177931           365                Fine particles (PM 2.5)   
9     742274           365                Fine particles (PM 2.5)   

          Measure Measure Info Geo Type Name  Geo Join ID  \
0  Number per km2       number         UHF42        409.0   
1  Number per km2       number         UHF42        209.0   
2            Mean       mcg/m3         

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names:
Renaming Columns: The columns were renamed for consistency. Spaces were replaced with underscores, and all letters were converted to lowercase using a list comprehension.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed

Checking and Correcting Data Types: Column data types were reviewed using dtypes, and any necessary conversions were made using pd.to_numeric() and pd.to_datetime().

3. Fix missing and invalid values in data.
Handling Missing Values was used: Missing values were filled out using fillna() for the data_value column, setting missing entries to 0 where appropriate.

4. Create new columns based on existing columns or calculations.
Creating New Columns: A new column, year, was extracted from the start_date column using dt.year.

5. Drop column(s) from your dataset.
Dropping Unnecessary Columns: The message column was dropped using drop(), as it did not add value to the analysis.

6. Drop a row(s) from your dataset.
Dropping Rows with Missing Values: Rows with any remaining missing values were removed using dropna().

7. Sort your data based on multiple variables.

Sorting Data: The dataset was sorted based on geo_place_name and time_period using sort_values().

8. Filter your data based on some condition.
Filtering Data: Data was filtered for a specific condition (e.g., the year 2015) using loc[].

9. Convert all the string values to upper or lower cases in one column.
String Manipulation: All string values in the geo_place_name column were converted to lowercase using str.lower().

10. Check whether numeric values are present in a given column of your dataframe.
Numeric Check: Ensured that all values in the data_value column were numeric using pd.to_numeric().

11. Group your dataset by one column, and get the mean, min, and max values by group.
  * Groupby()
  * agg() or .apply()

Grouping and Aggregation: The dataset was grouped by geo_place_name and year to calculate the mean, min, and max of data_value using groupby() and agg().

12. Group your dataset by two columns and then sort the aggregated results within the groups.
For this part the dataset was grouped by geo_place_name and year to calculate the mean, min, and max od data_value usign groupby() and agg() also sort_values() was used to group the two columns geo_place_name and year.

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

In [ ]:
# Rename columns for consistency
dataset.columns = [col.strip().replace(" ", "_").lower() for col in dataset.columns]

# Check the structure of the data
print("\nData Types:")
print(dataset.dtypes)

# Convert columns to appropriate data types
dataset['unique_id'] = pd.to_numeric(dataset['unique_id'], errors='coerce')
dataset['indicator_id'] = pd.to_numeric(dataset['indicator_id'], errors='coerce')
dataset['geo_join_id'] = pd.to_numeric(dataset['geo_join_id'], errors='coerce')
dataset['data_value'] = pd.to_numeric(dataset['data_value'], errors='coerce')
dataset['start_date'] = pd.to_datetime(dataset['start_date'], errors='coerce')

# Fix missing values (example: filling NaN with 0 or other appropriate values)
dataset.fillna({'data_value': 0}, inplace=True)

# Create new column based on existing columns
dataset['year'] = dataset['start_date'].dt.year

# Drop unnecessary columns
dataset.drop(columns=['message'], inplace=True)

# Drop rows with any missing values
dataset.dropna(inplace=True)

# Sort data based on multiple variables
sorted_data = dataset.sort_values(by=['geo_place_name', 'time_period'])

# Filter data based on some condition (example: filter for year 2015)
filtered_data = dataset[dataset['year'] == 2015]

# Convert string values to lower case in one column
dataset['geo_place_name'] = dataset['geo_place_name'].str.lower()

# Check for numeric values in a column
numeric_check = pd.to_numeric(dataset['data_value'], errors='coerce')
print("\nNumeric Check in 'data_value' Column:")
print(numeric_check.notnull().all())

# Group data by one column and get mean, min, and max values
grouped_data = dataset.groupby('geo_place_name')['data_value'].agg(['mean', 'min', 'max'])

# Group data by two columns and sort within groups
grouped_sorted_data = dataset.groupby(['geo_place_name', 'year'])['data_value'].agg(['mean', 'min', 'max']).sort_values(by=['geo_place_name', 'year'])

# Display the results
print("\nRenamed Columns and Cleaned Data:")
print(dataset.head(10))

print("\nGrouped Data by 'geo_place_name':")
print(grouped_data.head(10))

print("\nGrouped and Sorted Data by 'geo_place_name' and 'year':")
print(grouped_sorted_data.head(10))


Data Types:
unique_id           int64
indicator_id        int64
name               object
measure            object
measure_info       object
geo_type_name      object
geo_join_id       float64
geo_place_name     object
time_period        object
start_date         object
data_value        float64
message           float64
dtype: object

Numeric Check in 'data_value' Column:
True

Renamed Columns and Cleaned Data:
   unique_id  indicator_id                                   name  \
0     179772           640  Boiler Emissions- Total SO2 Emissions   
1     179785           640  Boiler Emissions- Total SO2 Emissions   
2     178540           365                Fine particles (PM 2.5)   
3     178561           365                Fine particles (PM 2.5)   
4     823217           365                Fine particles (PM 2.5)   
5     177910           365                Fine particles (PM 2.5)   
6     177952           365                Fine particles (PM 2.5)   
7     177973           365    

# Conclusions  

The analysis of the air quality dataset revealed significant geographical and temporal variations in pollutant levels, emphasizing the importance of monitoring and addressing air quality issues. Key observations included differences in SO2 emissions and PM 2.5 levels across various regions and years, with certain areas showing higher pollution levels. Data cleaning and preparation steps, such as renaming columns, handling missing values, and converting data types, ensured the integrity of the analysis and provided a reliable foundation for further exploration.

Future analysis could delve deeper into time series analysis to identify long-term trends and seasonal variations in air quality. Geospatial analysis could visualize pollution hotspots, while correlation studies might uncover relationships between pollutants and factors like population density or industrial activity. Predictive modeling could help forecast future air quality levels, and comparative analysis could identify patterns and causes for variations across different regions. These additional analyses would offer more comprehensive insights into air quality trends and inform public health and environmental policies.